accounts/abi/bind/bind.go

```diff
@@ -270,46 +398,73 @@ func Bind(types []string, abis []string, bytecodes []string, fsigs []map[string]
 		_, ok := isLib[types[i]]
 		contracts[types[i]].Library = ok
 	}
-	// Generate the contract template data content and render it
-	data := &tmplData{
-		Package:   pkg,
-		Contracts: contracts,
-		Libraries: libs,
-		Structs:   structs,
-	}
-	buffer := new(bytes.Buffer)
 
-	funcs := map[string]interface{}{
-		"bindtype":      bindType[lang],
-		"bindtopictype": bindTopicType[lang],
-		"namedtype":     namedType[lang],
-		"capitalise":    capitalise,
-		"decapitalise":  decapitalise,
-	}
-	tmpl := template.Must(template.New("").Funcs(funcs).Parse(tmplSource[lang]))
-	if err := tmpl.Execute(buffer, data); err != nil {
-		return "", err
-	}
-	// For Go bindings pass the code through gofmt to clean it up
-	if lang == LangGo {
-		code, err := format.Source(buffer.Bytes())
-		if err != nil {
-			return "", fmt.Errorf("%v\n%s", err, buffer)
+	// compute the full set of libraries that each contract depends on.
+	for _, contract := range contracts {
```

Comment:
Recursively finding the dependencies sounds very much like something that should exist as a standalone method, so that it can be properly unit-tested, and not just inlined within a huge for-loop. Please extract

Reply:
I like this because in order to extract the logic into a function that can be unit tested, I have to abstract the details of contract deployment, so that I can mock it.

so `LinkAndDeploy` would become something like:

```
type ContractDeployer interface {
    DeployContract(input []byte, deployer []byte) (common.Address, error)
}

func LinkAndDeploy(deployer ContractDeployer, deployParams DeploymentParams) (res *DeploymentResult, err error) {
 ...
}
```

It will end up making the interface for `LinkAndDeploy` cleaner.

Reply:
Ping?

Reply:
I've changed this part so that it uses the `depTreeBuilder` to compute for each contract/library, the entire set of dependencies (`MetaData` objects) that it requires.  So this improves code reuse a lot:  we share the same path for computing the dependency set here, and computing it before linking-and-deploying.

I still want to further refactor `bind` because it's huge and messy (as you've mentioned elsewhere).